### ISSUE TITLE
ClassCastException when using MockDataProvider with Oracle ArrayRecord types

### ISSUE THREAD

**Issue by lukaseder** [jOOQ/jOOQ#10704]
The following test case shows that currently, Oracle's `ArrayRecord` types cannot be placed in a `MockDataProvider`:

```java
static class StringTypeArrayRecord extends ArrayRecordImpl<String> {
    StringTypeArrayRecord(String... array) {
        super(null, "STRINGS", VARCHAR(10));
        addAll(Arrays.asList(array));
    }
}

static class IntegerTypeArrayRecord extends ArrayRecordImpl<Integer> {
    IntegerTypeArrayRecord(Integer... array) {
        super(null, "NUMBERS", INTEGER);
        addAll(Arrays.asList(array));
    }
}

static class ObjectType extends UDTImpl<ObjectTypeRecord> {
    ObjectType() {
        super("U_BOOK_TYPE", null);
    }

    static final ObjectType OBJECT_TYPE = new ObjectType();

    static final UDTField<ObjectTypeRecord, Integer> A = createField(DSL.name("A"), INTEGER, OBJECT_TYPE, "");
    static final UDTField<ObjectTypeRecord, String>  B = createField(DSL.name("B"), VARCHAR(400), OBJECT_TYPE, "");

    @Override
    public Class<ObjectTypeRecord> getRecordType() {
        return ObjectTypeRecord.class;
    };
}

static class ObjectTypeRecord extends org.jooq.impl.UDTRecordImpl<ObjectTypeRecord> {
    public ObjectTypeRecord() {
        super(OBJECT_TYPE);
    }
    public ObjectTypeRecord(Integer a, String b) {
        super(OBJECT_TYPE);
        set(0, a);
        set(1, b);
    }
}

static class ObjectTypeArrayRecord extends ArrayRecordImpl<ObjectTypeRecord> {
    ObjectTypeArrayRecord(ObjectTypeRecord... array) {
        super(null, "OBJECTS", OBJECT_TYPE.getDataType());
        addAll(Arrays.asList(array));
    }
}

final Field<StringTypeArrayRecord>  oracleStrings = field("STRINGS", VARCHAR(10).asArrayDataType(StringTypeArrayRecord.class));
final Field<IntegerTypeArrayRecord> oracleNumbers = field("INTEGERS", INTEGER.asArrayDataType(IntegerTypeArrayRecord.class));
final Field<ObjectTypeArrayRecord>  oracleObjects = field("OBJECTS", OBJECT_TYPE.getDataType().asArrayDataType(ObjectTypeArrayRecord.class));

class OracleArrayResult extends AbstractResult {

    final int              length;

    OracleArrayResult(int length) {
        this.length = length;
    }

    @Override
    public MockResult[] execute(MockExecuteContext ctx) throws SQLException {
        Result<Record3<StringTypeArrayRecord, IntegerTypeArrayRecord, ObjectTypeArrayRecord>> result =
            create.newResult(oracleStrings, oracleNumbers, oracleObjects);

        Record3<StringTypeArrayRecord, IntegerTypeArrayRecord, ObjectTypeArrayRecord> empty =
            create.newRecord(oracleStrings, oracleNumbers, oracleObjects);
        result.add(empty);

        for (int i = 0; i < length; i++) {
            Record3<StringTypeArrayRecord, IntegerTypeArrayRecord, ObjectTypeArrayRecord> record =
                create.newRecord(oracleStrings, oracleNumbers, oracleObjects);

            record.set(oracleStrings, new StringTypeArrayRecord(nCopies(i, "" + i).toArray(String[]::new)));
            record.set(oracleNumbers, new IntegerTypeArrayRecord(nCopies(i, i).toArray(Integer[]::new)));
            record.set(oracleObjects, new ObjectTypeArrayRecord(Seq.of(i).cycle(i).map(n -> new ObjectTypeRecord(n, "" + n)).toArray(ObjectTypeRecord[]::new)));

            result.add(record);
        }

        return new MockResult[] {
            new MockResult(0, result)
        };
    }
}

@Test
public void testOracleArrays() throws SQLException {
    DSLContext e = DSL.using(new MockConnection(new OracleArrayResult(3)), ORACLE);

    ResultSet r = e.select(oracleStrings, oracleNumbers, oracleObjects).fetchResultSet();
    assertEquals(3, r.getMetaData().getColumnCount());

    assertTrue(r.next());
    assertNull(r.getArray("STRINGS"));
    assertNull(r.getArray("INTEGERS"));
    assertNull(r.getArray("OBJECTS"));

    assertTrue(r.next());
    assertArrayEquals(new String[0], (String[]) r.getArray("STRINGS").getArray());
    assertArrayEquals(new Integer[0], (Integer[]) r.getArray("INTEGERS").getArray());
    // TODO: "OBJECTS" tests

    assertTrue(r.next());
    assertArrayEquals(new String[] { "1" }, (String[]) r.getArray("STRINGS").getArray());
    assertArrayEquals(new Integer[] { 1 }, (Integer[]) r.getArray("INTEGERS").getArray());

    assertTrue(r.next());
    assertArrayEquals(new String[] { "2", "2" }, (String[]) r.getArray("STRINGS").getArray());
    assertArrayEquals(new Integer[] { 2, 2 }, (Integer[]) r.getArray("INTEGERS").getArray());

    assertFalse(r.next());
}
```

The exception is:

```
java.lang.ClassCastException: class [Ljava.sql.Array; cannot be cast to class java.sql.Array ([Ljava.sql.Array; and java.sql.Array are in module java.sql of loader 'platform')
	at org.jooq.tools.jdbc.MockResultSet.getArray(MockResultSet.java:782)
	at org.jooq.tools.jdbc.DefaultResultSet.getArray(DefaultResultSet.java:713)
	at org.jooq.impl.CursorImpl$CursorResultSet.getArray(CursorImpl.java:623)
	at org.jooq.test.MockTest.testOracleArrays(MockTest.java:750)
```

**Comment by lukaseder** [comment]
This is due to an incomplete implementation of https://github.com/jOOQ/jOOQ/issues/3443, which was implemented for conversion between `Object[]` -> `java.sql.Array` only, not `Collection` -> `java.sql.Array`

**Comment by lukaseder** [comment]
The logic is already in place, but it seems there's a regression introduced by https://github.com/jOOQ/jOOQ/issues/3062

**Comment by lukaseder** [comment]
Fixed in jOOQ 3.14.0. Backport scheduled for 3.13.5 (#10706)
